File: protocol-generator/ocaml/src/base58.ml
(* Base58 (Bitcoin alphabet) — hand-rolled, used for peer-id formatting.
   Standard byte-array long-division; no bignum dependency needed. Leading
   zero bytes map to leading '1' characters. *)

let alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

let encode (input : string) : string =
  let len = String.length input in
  (* count leading zero bytes *)
  let zeros = ref 0 in
  while !zeros < len && input.[!zeros] = '\000' do incr zeros done;
  (* base256 → base58 via repeated division, accumulated big-endian *)
  let size = (len * 138 / 100) + 1 in
  let b58 = Bytes.make size '\000' in
  let high = ref (size - 1) in
  for i = 0 to len - 1 do
    let carry = ref (Char.code input.[i]) in
    let j = ref (size - 1) in
    while !j > !high || !carry <> 0 do
      carry := !carry + (256 * Char.code (Bytes.get b58 !j));
      Bytes.set b58 !j (Char.chr (!carry mod 58));
      carry := !carry / 58;
      decr j
    done;
    high := !j
  done;
  (* skip leading zero "digits" in the b58 buffer *)
  let start = ref 0 in
  while !start < size && Bytes.get b58 !start = '\000' do incr start done;
  let buf = Buffer.create (size + !zeros) in
  for _ = 1 to !zeros do Buffer.add_char buf '1' done;
  for k = !start to size - 1 do
    Buffer.add_char buf alphabet.[Char.code (Bytes.get b58 k)]
  done;
  Buffer.contents buf

(* index of each alphabet char, -1 if not a base58 char *)
let values =
  let v = Array.make 128 (-1) in
  String.iteri (fun i c -> v.(Char.code c) <- i) alphabet;
  v

let decode (s : string) : string =
  let len = String.length s in
  let ones = ref 0 in
  while !ones < len && s.[!ones] = '1' do incr ones done;
  let size = (len * 733 / 1000) + 1 in (* log(58)/log(256) ≈ 0.733 *)
  let b256 = Bytes.make size '\000' in
  let high = ref (size - 1) in
  for i = 0 to len - 1 do
    let c = Char.code s.[i] in
    let d = if c < 128 then values.(c) else -1 in
    if d < 0 then invalid_arg "Base58.decode: invalid character";
    let carry = ref d in
    let j = ref (size - 1) in
    while !j > !high || !carry <> 0 do
      carry := !carry + (58 * Char.code (Bytes.get b256 !j));
      Bytes.set b256 !j (Char.chr (!carry land 0xff));
      carry := !carry lsr 8;
      decr j
    done;
    high := !j
  done;
  let start = ref 0 in
  while !start < size && Bytes.get b256 !start = '\000' do incr start done;
  let buf = Buffer.create (size + !ones) in
  for _ = 1 to !ones do Buffer.add_char buf '\000' done;
  Buffer.add_string buf (Bytes.sub_string b256 !start (size - !start));
  Buffer.contents buf

File: protocol-generator/nim/src/base58.nim
## Base58 (Bitcoin alphabet) — hand-rolled, used for peer-id formatting.
## Standard byte-array long-division; no bignum dependency (profile
## [codec].base58_library = hand-rolled). Leading zero bytes map to leading '1'.
##
## SPDX-License-Identifier: Apache-2.0

const Alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

proc base58Encode*(input: openArray[byte]): string =
  let n = input.len
  var zeros = 0
  while zeros < n and input[zeros] == 0'u8: inc zeros

  let size = (n * 138 div 100) + 1
  var b58 = newSeq[byte](size)  # zero-initialised

  var high = size - 1
  for i in 0 ..< n:
    var carry = int(input[i])
    var j = size - 1
    while j > high or carry != 0:
      carry += 256 * int(b58[j])
      b58[j] = byte(carry mod 58)
      carry = carry div 58
      if j == 0: break
      dec j
    high = j

  var start = 0
  while start < size and b58[start] == 0'u8: inc start

  result = newStringOfCap(zeros + (size - start))
  for _ in 0 ..< zeros: result.add '1'
  for k in start ..< size: result.add Alphabet[int(b58[k])]

proc b58Value(c: char): int =
  case c
  of '1'..'9': int(c) - int('1')
  of 'A'..'H': int(c) - int('A') + 9
  of 'J'..'N': int(c) - int('J') + 17
  of 'P'..'Z': int(c) - int('P') + 22
  of 'a'..'k': int(c) - int('a') + 33
  of 'm'..'z': int(c) - int('m') + 44
  else: -1

proc base58Decode*(s: string): seq[byte] =
  let n = s.len
  var ones = 0
  while ones < n and s[ones] == '1': inc ones

  let size = (n * 733 div 1000) + 1  # log(58)/log(256) ~= 0.733
  var b256 = newSeq[byte](size)

  var high = size - 1
  for i in 0 ..< n:
    let d = b58Value(s[i])
    if d < 0: raise newException(ValueError, "invalid base58 character")
    var carry = d
    var j = size - 1
    while j > high or carry != 0:
      carry += 58 * int(b256[j])
      b256[j] = byte(carry and 0xff)
      carry = carry shr 8
      if j == 0: break
      dec j
    high = j

  var start = 0
  while start < size and b256[start] == 0'u8: inc start

  result = newSeq[byte](ones + (size - start))
  for k in start ..< size: result[ones + (k - start)] = b256[k]
